google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/AppEngineAction.java

```diff
@@ -91,14 +94,30 @@ protected void executeProcess(
 
     consoleLogLn("Executing: " + commandLine.getCommandLineString());
 
-    final Process process = commandLine.createProcess();
-    final ProcessHandler processHandler = new OSProcessHandler(process,
-        commandLine.getCommandLineString());
+    Process process = commandLine.createProcess();
+    processHandler = new OSProcessHandler(process, commandLine.getCommandLineString());
     loggingHandler.attachToProcess(processHandler);
     processHandler.addProcessListener(listener);
+
+    // mark process as completed by setting handler to null when terminated
+    processHandler.addProcessListener(new ProcessAdapter() {
+      @Override
+      public void processTerminated(ProcessEvent event) {
+        processHandler = null;
+      }
+    });
+
     processHandler.startNotify();
   }
 
+  protected void cancel() {
+    // kill any executing process for the action
+    if (processHandler != null && processHandler.getProcess() != null) {
+      cancelled = true;
+      processHandler.getProcess().destroy();
+    }
```

Comment:
Hmmmm. I don't think I get it. Not sure if I am missing something. Here's what I thought:

The registered listener that sets `processHandler = null` will be run by a different thread when the gcloud deploy command starts. So that listener can set `processHandler` to null anytime, regardless of `disconnect()` is synchronized or an action can be executed or canceled more than once. There is nothing preventing that listener from setting the handler to null. The listener will do this when the gcloud process exits, and that can coincide with when `disconnect()` is called. I feel like we need to synchronize `processHandler`, since two threads can access the handler concurrently.

What do you think?